keps/sig-storage/3314-csi-changed-block-tracking/README.md

```diff
@@ -0,0 +1,1689 @@
+* Provide an API to retrieve the data blocks of a snapshot.
+  > *It is assumed that a snapshot's data blocks can be retrieved by creating a
+    PersistentVolume for the snapshot, launching a pod with this volume
+    attached in `Block` volume mode, and then reading the individual
+    blocks from the raw block device.*
+
+* Support of file changed list tracking for network file shares is not
+addressed by this proposal.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal extends the CSI specification with a new, optional,
+[CSI SnapshotMetadata gRPC service](#the-csi-snapshotmetadata-service-api),
+that is used by Kubernetes to retrieve metadata on the allocated blocks of a
+single snapshot, or the changed blocks between a pair of snapshots of the same
+block volume.
+
+A [Kubernetes SnapshotMetadata gRPC service](#the-kubernetes-snapshotmetadata-service-api)
+is an API that is used by a Kubernetes backup application client to retrieve 
+snapshot metadata. This API is implemented by the the community provided
+[external-snapshot-metadata sidecar](#the-external-snapshot-metadata-sidecar),
+which must be deployed by a CSI driver.
+A Kubernetes backup application will retrieve snapshot metadata through a
+TLS gRPC connection to such a service.
+This direct connection results in a minimal load on the Kubernetes API server,
+unrelated to the amount of metadata transferred or the sizes of the volumes
+and snapshots involved.
+
+The [external-snapshot-metadata sidecar](#the-external-snapshot-metadata-sidecar)
+communicates over a private UNIX domain socket with the CSI driver's
+implementation of the [CSI SnapshotMetadata](#the-csi-snapshotmetadata-service-api)
+gRPC service.
+The CSI driver service only handles the retrieval of the metadata requested;
+the sidecar is responsible for validating the Kubernetes authentication token,
+authorizing the backup application, validating the parameters of the RPC calls
+and fetching the provisioner secrets needed to complete the request.
+The sidecar forwards the RPC call to the CSI driver service over the UNIX
+domain socket, after translating Kubernetes object names into SP object names,
+and re-streams the results back to its client.
+
+A CSI driver advertises the existence of the community sidecar's
+[Kubernetes SnapshotMetadata](#the-kubernetes-snapshotmetadata-service-api) gRPC service
+to Kubernetes backup applications
+by creating a [SnapshotMetadataService CR](#snapshot-metadata-service-custom-resource)
+that contains the service's TCP endpoint address, CA certificate and
+an audience string needed for token authentication.
```

Comment:
Let's follow up offline on this, but we may want to enforce some syntax/restrictions on the allowed audience strings. The primary goal is to prevent reusing some reserved audience for kube-apiserver. cc @liggitt